File: packages/java/src/main/java/com/peruspatial/crs/ToWGS84.java
package com.peruspatial.crs;

import com.peruspatial.data.CrsData;

public final class ToWGS84 {
  private ToWGS84() {}

  public static Wgs84Result toWGS84(double easting, double northing, int zone) {
    double centralMeridian = (zone - 1) * 6 - 180 + 3;
    for (CrsData.Zone z : CrsData.ZONES) {
      if (z.zone() == zone) {
        centralMeridian = z.centralMeridian();
        break;
      }
    }

    double a = CrsData.A;
    double f = CrsData.F;
    double k0 = CrsData.K0;
    double e2 = f * (2 - f);
    double ep2 = e2 / (1 - e2);

    double x = easting - CrsData.FALSE_EASTING;
    double y = northing - CrsData.FALSE_NORTHING_SOUTH;

    double M = y / k0;
    double mu = M / (a * (1 - e2 / 4 - 3 * e2 * e2 / 64 - 5 * e2 * e2 * e2 / 256));

    double e1 = (1 - Math.sqrt(1 - e2)) / (1 + Math.sqrt(1 - e2));

    double phi1 = mu
        + (3 * e1 / 2 - 27 * e1 * e1 * e1 / 32) * Math.sin(2 * mu)
        + (21 * e1 * e1 / 16 - 55 * e1 * e1 * e1 * e1 / 32) * Math.sin(4 * mu)
        + (151 * e1 * e1 * e1 / 96) * Math.sin(6 * mu)
        + (1097 * e1 * e1 * e1 * e1 / 512) * Math.sin(8 * mu);

    double sinPhi1 = Math.sin(phi1);
    double cosPhi1 = Math.cos(phi1);
    double tanPhi1 = Math.tan(phi1);

    double N1 = a / Math.sqrt(1 - e2 * sinPhi1 * sinPhi1);
    double T1 = tanPhi1 * tanPhi1;
    double C1 = ep2 * cosPhi1 * cosPhi1;
    double R1 = a * (1 - e2) / Math.pow(1 - e2 * sinPhi1 * sinPhi1, 1.5);
    double D = x / (N1 * k0);

    double phi = phi1 - (N1 * tanPhi1 / R1) * (D * D / 2
        - (5 + 3 * T1 + 10 * C1 - 4 * C1 * C1 - 9 * ep2) * Math.pow(D, 4) / 24
        + (61 + 90 * T1 + 298 * C1 + 45 * T1 * T1 - 252 * ep2 - 3 * C1 * C1) * Math.pow(D, 6) / 720);

    double lambda = Math.toRadians(centralMeridian) + (D
        - (1 + 2 * T1 + C1) * Math.pow(D, 3) / 6
        + (5 - 2 * C1 + 28 * T1 - 3 * C1 * C1 + 8 * ep2 + 24 * T1 * T1) * Math.pow(D, 5) / 120) / cosPhi1;

    return new Wgs84Result(Math.toDegrees(phi), Math.toDegrees(lambda));
  }
}

File: packages/rust/src/crs/to_wgs84.rs
use super::to_utm::central_meridian_for;
use super::types::Wgs84;
use crate::data::crs::{A, F, FALSE_EASTING, FALSE_NORTHING_SOUTH, K0};

pub fn to_wgs84(easting: f64, northing: f64, zone: i32) -> Wgs84 {
    let e2 = F * (2.0 - F);
    let ep2 = e2 / (1.0 - e2);
    let central_meridian = central_meridian_for(zone);

    let x = easting - FALSE_EASTING;
    let y = northing - FALSE_NORTHING_SOUTH;

    let m = y / K0;
    let mu = m
        / (A * (1.0 - e2 / 4.0 - 3.0 * e2 * e2 / 64.0 - 5.0 * e2 * e2 * e2 / 256.0));

    let e1 = (1.0 - (1.0 - e2).sqrt()) / (1.0 + (1.0 - e2).sqrt());

    let phi1 = mu
        + (3.0 * e1 / 2.0 - 27.0 * e1.powi(3) / 32.0) * (2.0 * mu).sin()
        + (21.0 * e1 * e1 / 16.0 - 55.0 * e1.powi(4) / 32.0) * (4.0 * mu).sin()
        + (151.0 * e1.powi(3) / 96.0) * (6.0 * mu).sin()
        + (1097.0 * e1.powi(4) / 512.0) * (8.0 * mu).sin();

    let sin_phi1 = phi1.sin();
    let cos_phi1 = phi1.cos();
    let tan_phi1 = phi1.tan();

    let n1 = A / (1.0 - e2 * sin_phi1 * sin_phi1).sqrt();
    let t1 = tan_phi1 * tan_phi1;
    let c1 = ep2 * cos_phi1 * cos_phi1;
    let r1 = A * (1.0 - e2) / (1.0 - e2 * sin_phi1 * sin_phi1).powf(1.5);
    let d = x / (n1 * K0);

    let lat = phi1
        - (n1 * tan_phi1 / r1)
            * (d * d / 2.0
                - (5.0 + 3.0 * t1 + 10.0 * c1 - 4.0 * c1 * c1 - 9.0 * ep2) * d.powi(4) / 24.0
                + (61.0 + 90.0 * t1 + 298.0 * c1 + 45.0 * t1 * t1 - 252.0 * ep2 - 3.0 * c1 * c1)
                    * d.powi(6)
                    / 720.0);

    let lng_rad = (d - (1.0 + 2.0 * t1 + c1) * d.powi(3) / 6.0
        + (5.0 - 2.0 * c1 + 28.0 * t1 - 3.0 * c1 * c1 + 8.0 * ep2 + 24.0 * t1 * t1) * d.powi(5)
            / 120.0)
        / cos_phi1;

    Wgs84 {
        lat: lat.to_degrees(),
        lng: central_meridian + lng_rad.to_degrees(),
    }
}
